### ISSUE TITLE
Problems with inference in unions of tuples

### ISSUE THREAD

**Issue by saabi** [Microsoft/TypeScript#21949]
Search Terms: tuples unions inference

In the following example, for constant `f`, the inferred type should be `number | undefined` (or even narrowed down to `undefined` in this particularly obvious case) but instead it's being incorrectly inferred as `string`.

I tried this on 2.8.0-dev.20180214 as well and it's still showing the same behavior.

```ts
type a = [string] | [string, number]

const b: a = ['bye', 1];    // accepted by the compiler
const c: a = ['hi'];        // accepted by the compiler
const d: a = ['hi', 'bye']  // correctly throws an error

const e = b[1];  // inferred type of e is number
const f = c[1];  // inferred type of f is string (!?) should be: number | undefined
```

Here's a [playground link](https://www.typescriptlang.org/play/#src=type%20a%20%3D%20%5Bstring%5D%20%7C%20%5Bstring%2C%20number%5D%0D%0A%0D%0Aconst%20b%3A%20a%20%3D%20%5B'bye'%2C%201%5D%3B%20%20%20%20%2F%2F%20accepted%20by%20the%20compiler%0D%0Aconst%20c%3A%20a%20%3D%20%5B'hi'%5D%3B%20%20%20%20%20%20%20%20%2F%2F%20accepted%20by%20the%20compiler%0D%0Aconst%20d%3A%20a%20%3D%20%5B'hi'%2C%20'bye'%5D%20%20%2F%2F%20correctly%20throws%20an%20error%0D%0A%0D%0Aconst%20e%20%3D%20b%5B1%5D%3B%20%20%2F%2F%20inferred%20type%20of%20e%20is%20number%0D%0Aconst%20f%20%3D%20c%5B1%5D%3B%20%20%2F%2F%20inferred%20type%20of%20f%20is%20string%20(!%3F)%20should%20be%3A%20number%20%7C%20undefined%0D%0A)

**Comment by saabi** [comment]
Ok, after investigating the playground example a little bit more I get what's going on.

Type narrowing is being too aggressive (to my expectations anyway) and the compiler is narrowing `c` down to a simple `string[]`. The rest of the behavior is to be expected after that.

This is not an implementation bug, but I think it's arguable whether it's a design bug, as it kind of makes futile the intent of having the tuple union in the first place. It probably depends on what that intention is in the first place, but there certainly is no way of specifying that in the definition.

The only fix (... I can think of so far...) is to cast at every access, which is kind of annoying. If I have to cast everywhere I use such a variable it's like not casting anywhere since the possibility of introducing a casting error somewhere defeats the purpose of typing.

**Comment by RyanCavanaugh** [comment]
Basically a duplicate of #5203. We know (from analyzing the code) that `c` is `[string]` (you can see this by writing `['hi'] as any` instead). But if you try to OOB a tuple we just defer to a union of its element types because, well, we do (I don't have a good defense of this behavior).

**Comment by jcalz** [comment]
Depending on your use case, you could change `a` to be

```ts
type a = {0: string, 1?: never, length: 1}  | {0: string, 1: number, length: 2}
```
or
```ts
type a = ([string] & {1?: never})| [string, number]
```

**Comment by saabi** [comment]
Continuing with my argument from my last comment:

> If narrowing is unavoidable, then it should probably narrow to '[string]' instead of 'string[]'

I tried it on the [playground](https://www.typescriptlang.org/play/#src=type%20a%20%3D%20%5Bstring%5D%20%7C%20%5Bstring%2C%20number%5D%0D%0A%0D%0Aconst%20b%3A%20a%20%3D%20%5B'bye'%2C%201%5D%3B%20%20%20%20%2F%2F%20accepted%20by%20the%20compiler%0D%0Aconst%20c%3A%20a%20%3D%20%5B'hi'%5D%3B%20%20%20%20%20%20%20%20%2F%2F%20accepted%20by%20the%20compiler%0D%0Aconst%20d%3A%20a%20%3D%20%5B'hi'%2C%20'bye'%5D%20%20%2F%2F%20correctly%20throws%20an%20error%0D%0A%0D%0Aconst%20e%20%3D%20b%5B1%5D%3B%20%20%2F%2F%20inferred%20type%20of%20e%20is%20number%0D%0Aconst%20f%20%3D%20c%5B1%5D%3B%20%20%2F%2F%20inferred%20type%20of%20f%20is%20string%20(!%3F)%20should%20be%3A%20number%20%7C%20undefined%0D%0A%0D%0Atype%20g%20%3D%20%5Bstring%5D%3B%0D%0A%0D%0Aconst%20h%20%3D%20%5B%22I'm%20back!%22%5D%3B%20%20%20%20%2F%2F%20inferred%20as%20string%5B%5D%0D%0Aconst%20i%3A%20g%20%3D%20%5B%22I'm%20back!%22%5D%3B%0D%0A%0D%0Aconst%20j%20%3D%20h%5B1%5D%3B%20%2F%2F%20j%20is%20inferred%20as%20string.%20Correct%0D%0Aconst%20k%20%3D%20i%5B1%5D%3B%20%2F%2F%20k%20is%20inferres%20as%20string.%20(mmmh...%3F)%0D%0A%2F%2F%20Shouldn't%20it%20be%20undefined%20after%20typescript%202.7%3F)

```ts
type g = [string];

const h = ["I'm back!"];    // inferred as string[]
const i: g = ["I'm back!"];

const j = h[1]; // j is inferred as string. Correct
const k = i[1]; // k is inferres as string. (mmmh...?)
// Shouldn't it be undefined after typescript 2.7?
```

**Comment by saabi** [comment]
@jcalz You're right, that would work. 

But after investigating on the playground a bit more (see my previous comments) and after the introduction of fixed length tuples in 2.7, I think there might actually be an implementation bug (not the one imagined in the OP)
